Changes in module 'nf-core/bcftools/annotate'
--- modules/nf-core/bcftools/annotate/main.nf
+++ modules/nf-core/bcftools/annotate/main.nf
@@ -14,7 +14,7 @@
     tuple val(meta), path("${prefix}.${extension}"), emit: vcf
     tuple val(meta), path("${prefix}.${extension}.tbi"), emit: tbi, optional: true
     tuple val(meta), path("${prefix}.${extension}.csi"), emit: csi, optional: true
-    tuple val("${task.process}"), val('bcftools'), eval("bcftools --version | sed '1!d; s/^.*bcftools //'"), topic: versions, emit: versions_bcftools
+    path "versions.yml", emit: versions
 
     when:
     task.ext.when == null || task.ext.when
@@ -51,6 +51,11 @@
         --output ${prefix}.${extension} \\
         --threads ${task.cpus} \\
         ${input}
+
+    cat <<-END_VERSIONS > versions.yml
+    "${task.process}":
+        bcftools: \$(bcftools --version 2>&1 | head -n1 | sed 's/^.*bcftools //; s/ .*\$//')
+    END_VERSIONS
     """
 
     stub:
@@ -77,5 +82,10 @@
     """
     ${create_cmd} ${prefix}.${extension}
     ${create_index}
+
+    cat <<-END_VERSIONS > versions.yml
+    "${task.process}":
+        bcftools: \$(bcftools --version 2>&1 | head -n1 | sed 's/^.*bcftools //; s/ .*\$//')
+    END_VERSIONS
     """
 }

************************************************************
